### PR TITLE
Save and restore all opening eaf browser buffers and pdf files.

### PR DESCRIPTION
**Pull request by lhpfvs**
Add pdf files to the saved Emacs session along with browser buffers, so both EAF Browser buffers and pdf files opened in the previous session can be restored.

### CODE REVIEW

File: eaf.el [hunk 1]
```diff
@@ -932,13 +931,20 @@ This should be used after setting `eaf-browser-continue-where-left-off' to t."
                                 (split-string (buffer-string) "\n" t))))
         (if (epc:live-p eaf-epc-process)
             (dolist (url browser-url-list)
-              (eaf-open-browser url))
+              (if (string-prefix-p "http" url)
```
Comment (manateelazycat) [inline eaf.el:934]: string-prefix-p 应该用 eaf-is-valid-web-url 替换

File: eaf.el [hunk 2]
```diff
@@ -932,13 +931,20 @@ This should be used after setting `eaf-browser-continue-where-left-off' to t."
...
+                  (eaf-open-browser url)
+                (eaf-open url)))
           (dolist (url browser-url-list)
-            (push `(,url "browser" "") eaf--active-buffers))
-          (when eaf--active-buffers (eaf-open-browser (nth 0 (car eaf--active-buffers))))))
+            (if (string-prefix-p "http" url)
```
Comment (manateelazycat) [inline eaf.el:938]: 用 eaf-is-valid-web-url 替换

File: eaf.el [hunk 3]
```diff
@@ -932,13 +931,20 @@ This should be used after setting `eaf-browser-continue-where-left-off' to t."
...
+                (push `(,url "browser" "") eaf--active-buffers)
+              (push `(,url "pdf-viewer" "") eaf--active-buffers)))
+          (when eaf--active-buffers
+            (let ((url (nth 0 (car eaf--active-buffers))))
+              (if (string-prefix-p "http" url)
```
Comment (manateelazycat) [inline eaf.el:943]: 用 eaf-is-valid-web-url 替换

**Review by manateelazycat — CHANGES_REQUESTED** [review]

**Comment by manateelazycat** [comment]
@MatthewZMD 你也帮忙review一下，看看代码有没有问题？

File: eaf.el [hunk 4]
```diff
@@ -1458,22 +1464,24 @@ keybinding variable to eaf-app-binding-alist."
...
-    (when eaf-browser-continue-where-left-off
-      (let* ((browser-restore-file-path
+(defun eaf--save-session ()
+  "Save opening eaf browser tabs and pdf files"
+  (let* ((browser-restore-file-path
               (concat eaf-config-location
                       (file-name-as-directory "browser")
                       (file-name-as-directory "history")
-                      "restore.txt"))
-             (browser-urls ""))
```
Comment (manateelazycat) [inline eaf.el:1470]: 为什么要删除这两行代码呢？

**Review by manateelazycat — CHANGES_REQUESTED** [review]

File: eaf.el (hunk 4 above)
Comment (lhpfvs) [reply eaf.el:1475]: 抱歉，应该是不小心按错键了，我会注意review。

File: eaf.el [hunk 5]
```diff
@@ -1463,22 +1469,24 @@ keybinding variable to eaf-app-binding-alist."
        (eaf--kill-python-process))
      )))
 
-(defun eaf--monitor-emacs-kill ()
-  "Function monitoring when Emacs is killed."
-  (ignore-errors
-    (when eaf-browser-continue-where-left-off
```
Comment (manateelazycat) [inline eaf.el:1469]: eaf-browser-continue-where-left-off 这个选项为什么要去掉？ 不是每个人都希望再次打开EAF浏览器的时候都恢复上次浏览的页面。

**Review by manateelazycat — CHANGES_REQUESTED** [review]

File: eaf.el [hunk 6]
```diff
@@ -932,13 +931,20 @@ This should be used after setting `eaf-browser-continue-where-left-off' to t."
...
-              (eaf-open-browser url))
+              (if (eaf-is-valid-web-url url)
+                  (eaf-open-browser url)
+                (eaf-open url)))
           (dolist (url browser-url-list)
-            (push `(,url "browser" "") eaf--active-buffers))
-          (when eaf--active-buffers (eaf-open-browser (nth 0 (car eaf--active-buffers))))))
+            (if (eaf-is-valid-web-url url)
+                (push `(,url "browser" "") eaf--active-buffers)
+              (push `(,url "pdf-viewer" "") eaf--active-buffers)))
```
Comment (MatthewZMD) [inline eaf.el:940]: eaf-pdf-viewer shouldn't be hard-coded here. It is understandable that the Browser is a much more unique feature than other EAF Applications that work with files, but it is weird that only PDF viewers can be restored but not other things such as Video players etc.

Comment (MatthewZMD) [reply eaf.el:940]: Try to make it more general and work for other EAF applications.

**Pull request closed without merging**